src/client.ts

```diff
@@ -381,6 +381,12 @@ export interface ICreateClientOpts {
      * Default: false.
      */
     isVoipWithNoMediaAllowed?: boolean;
+
+    /**
+     * A logger to associate with this MatrixClient.
+     * Defaults to the build-in global logger.
+     */
+    logger?: PrefixedLogger;
```

Comment:
It is slightly dubious to use `PrefixesLogger` here since it inherits from a class outside of our control, we could use something custom if we wanted to strictly define / document the interface that is required?

Reply:
> Yeah. It might be better to just accept a `logPrefix: string`, and make MatrixClient responsible for initialising the `PrefixedLogger`?

Personally I’d actually like more flexibility than just a custom prefix. I’d also like to disable logs altogether or configure the logger to only output certain levels.

Reply:
> Personally I’d actually like more flexibility than just a custom prefix. I’d also like to disable logs altogether or configure the logger to only output certain levels.

I totally agree with that, though I think it's orthogonal to the changes proposed here.

One idea that I had is to wire up the logging to [`debug`](https://www.npmjs.com/package/debug), and then the logging can be controlled via environment variables.

Reply:
The goal of the patch for "us" (TB) is to be able to log with a totally different logger. So we can control where and what levels get logged.

Reply:
> The goal of the patch for "us" (TB) is to be able to log with a totally different logger. So we can control where and what levels get logged.

Yes, fair. https://phabricator.services.mozilla.com/D181226 is instructive here.

Ok fine, I'm convinced, let's leave it as `PrefixedLogger`!

Reply:
As long as we can access the whole logger object (rather than just setting a string) I’m all fine with a very strict interface. 🙃

Reply:
> Sorry, I deleted my comment again because I think we could do with a slightly tighter interface, too, it'd just not be cleanly in "userspace" anymore.

It feels fairly dirty to me to re-use `PrefixedLogger` -- for one the Thunderbid patch doesn't even implement it as a prefix, but uses it as a suffix:

```javascript
  withPrefix(prefix) {
    return new MatrixLogger(this.prefix + prefix, this._account);
  },
```

@freaktechnik What was your thought on a tighter interface? This would also probably be desirable since it would allow changing the underlying logging library sued by matrix-js-sdk *without* being a breaking change?

Reply:
I'm not too familiar with `loglevel`, but something like having a function that takes a prefix, a level and then the stuff to log might be more flexible (`(string prefix, LogLevel level, ...logArgs) => void`)? That said, if the logger should expose all a JS console can do, that is not a good interface at all. Then having more of a factory style function might be preferable, where you specify prefix and get a console to log with (`(prefix: string) => Console`). In general I'd think that the library shouldn't expect to control its own log level when an external logger is used (which from what I can tell is what loglevel provides over a plain JS console).

It might also be better to call it module, context or similar instead of prefix, since as observed, not all loggers might log it before the rest of the information, and instead use it for other purposes altogether.

Reply:
I think I can understand where the apathy for something losely defined is coming from. But in my opinion either defining a custom (tight) interface or just allowing log levels to be set via a method are perfectly fine here.
Log levels are kind of standardized as per https://datatracker.ietf.org/doc/html/rfc5424#section-6.2.1 (bottom of 6.2.1)...

The most popular logging libraries (like Pino or Winston) are making use of that as well: https://www.npmjs.com/search?q=keywords:logger -- alternatively we can also just go with one of those and be done with this.
Or yeah, go ahead and use [loglevel](https://www.npmjs.com/package/loglevel).